mathics/core/parser/README.md

```diff
@@ -2,82 +2,91 @@
-attribute. Consider trying to parse the code snippet `a + b + c`. There is
-some ambiguity in what order the operations should be applied. Should we use
-left associativity `Plus[Plus[a + b] + c]` or right associativity
-`Plus[a, Plus[b + c]]`?
+### Associativity
+
+In addition to precedences, operators in the Wolfram language have a associativity or grouping attribute. Consider parsing `a || b || c`. There is some ambiguity in what order the operations should be applied. Should we use left associativity `Or[Or[a, b], c]` or right associativity `Or[a, Or[b, c]]`?
 
 #### Flat associativity
 
-In fact the addition operator is usually considered fully associative, that is
-`Plus[a, b, c]`. In the Wolfram language this is referred to as 'flat'
-associativity because the expression tree is flat.
+There is an additional kind of associativity that treats _n_ repeated uses of a binary operator as that single operator applied to _n_ arguments. For example, `Plus[a, [Plus[b, c]]` is the same thing as `Plus[a, b, c]` (or `Plus[Plus[a, b], c]`). In the Wolfram language this is referred to as ['Flat'](https://reference.wolfram.com/language/ref/Flat.html) associativity. The expression tree for such a sequence flattened to the operator and its _n_ operands.
 
-Flat associativity is not a large issue in practice as the as flat operators
-can be first treated as left or right associative, whichever is more efficient,
-and flattened once the AST is constructed.
+Flat associativity is not an issue in parsing, since the flat operators can be first treated as left or right associative, whichever is more efficient, and flattened once the AST is constructed.
 
-One thing to be aware of is that parenthesis prevent the flattening of
-expressions at parse time, `a * (b * c)` parses as `Times[a, Times[b, c]]` but
-is flattened immediately afterwards.
+One thing to be aware of is that parenthesis prevent the flattening of expressions at parse time, `a * (b * c)` parses as `Times[a, Times[b, c]]` but is flattened immediately afterwards.
 
 #### Non-associativity
-In addition to left/right associativity some operators are non-associative.
-That is, an operator `?` is non-associative if `a ? b ? c` is not valid syntax.
+In addition to left/right associativity, some operators are non-associative. That is, an operator `?` is non-associative if `a ? b ? c` is not valid syntax.
 
-The only non associative operator in the Wolfram language is `PatternTest`.
+A non-associative operator in the Wolfram language is
+[`PatternTest`](https://reference.wolfram.com/language/ref/PatternTest.html?q=PatternTest).
 
 #### Implementation of binary operators
 
 The relevant code snippet for binary operators is:
 
 
 ```python
-def parse_binary(self, expr1, token, p):    # called with expr1
-    'expr <- expr1 BINARY expr2'
+def parse_binary(self, expr1, token, expr1_precedence: int) -> Node:
+    """
+    Implements grammar rule:
+	   expr ::= expr1 BINARY expr2
```

Comment:
Where was the `::=` meaning defined?

Reply:
I checked ANTR and BISON syntax and they both uses ":", so I'll change to this.

Reply:
The notation is OK, but it may be worth mentioning in this document.